pkg/filesystem/read_layer.go

```diff
@@ -0,0 +1,69 @@
+package filesystem
+
+import (
+	"io"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// A ReadLayer is a read only interface for use as backing data for file
+// implementations. It behaves as a read only data source which may still be
+// shrunk (or grown) with truncate calls. Reading past the logical size returns
+// zeroes, reading at negative offsets or truncating to negative sizes is an
+// error.
+type ReadLayer interface {
+	io.ReaderAt
+	Truncate(size int64) error
```

Comment:
I just realized: doesn't this need a GetNextRegionOffset() to make GetNextRegionOffset() against the pool backed file work correctly? Because whether a part of a file is sparse can now no longer be determined in isolation.